File: legacy_ios/LegacyApp/Sources/ConfirmFlutterViewController.m
#import "ConfirmFlutterViewController.h"
#import "BaseViewController.h"
#import "CompleteViewController.h"
#import <FlutterPluginRegistrant/GeneratedPluginRegistrant.h>

static NSString *const kChannelName = @"com.example.legacyapp/confirm";
static NSString *const kSubmitURLString = @"https://jsonplaceholder.typicode.com/posts";

@interface ConfirmFlutterViewController ()

@property (nonatomic, strong) FlutterEngine *ownedEngine;
@property (nonatomic, strong) FlutterMethodChannel *channel;

@end

@implementation ConfirmFlutterViewController

- (instancetype)init {
    // A fresh engine per visit (rather than one warm engine reused for the
    // app's lifetime) is deliberate here: Dart's main()/initState() only
    // ever runs once per engine, so a shared long-lived engine would only
    // ever answer getInitialData with whatever the form held the very
    // first time this screen was reached -- never the current values on a
    // second visit. A fresh engine keeps each visit correct at the cost of
    // the warm-start latency; a production app would instead keep one
    // engine alive and push fresh data into it explicitly rather than
    // relying on initState.
    FlutterEngine *engine = [[FlutterEngine alloc] initWithName:@"confirm_engine"];
    [GeneratedPluginRegistrant registerWithRegistry:engine];

    // engine.binaryMessenger only works once the engine is running
    // (registering a channel handler before run() crashes inside
    // FlutterBinaryMessengerRelay), so run() must happen first. Dart's
    // first frame (and the getInitialData call inside it) only fires once
    // this view controller's view actually loads, which happens later
    // when it's pushed on screen -- well after this synchronous init
    // method has finished setting up the channel handler below.
    [engine run];

    self = [super initWithEngine:engine nibName:nil bundle:nil];
    if (self) {
        _ownedEngine = engine;
        _channel = [FlutterMethodChannel methodChannelWithName:kChannelName
                                                binaryMessenger:engine.binaryMessenger];
        __weak typeof(self) weakSelf = self;
        [_channel setMethodCallHandler:^(FlutterMethodCall *call, FlutterResult result) {
            [weakSelf handleMethodCall:call result:result];
        }];
    }
    return self;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([call.method isEqualToString:@"getInitialData"]) {
        FormData *data = BaseViewController.sharedFormData;
        result(@{@"name": data.name, @"email": data.email, @"message": data.message});
    } else if ([call.method isEqualToString:@"confirmSubmit"]) {
        [self submitWithCompletion:^(BOOL success) {
            result(@(success));
        }];
    } else if ([call.method isEqualToString:@"goToComplete"]) {
        CompleteViewController *completeVC = [[CompleteViewController alloc] init];
        [self.navigationController pushViewController:completeVC animated:YES];
        result(nil);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)submitWithCompletion:(void (^)(BOOL success))completion {
    FormData *data = BaseViewController.sharedFormData;
    NSDictionary *body = @{
        @"name": data.name,
        @"email": data.email,
        @"message": data.message,
    };
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:body options:0 error:nil];

    NSURL *url = [NSURL URLWithString:kSubmitURLString];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    request.HTTPMethod = @"POST";
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    request.HTTPBody = jsonData;

    NSURLSessionDataTask *task = [[NSURLSession sharedSession] dataTaskWithRequest:request completionHandler:^(NSData * _Nullable respData, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        NSInteger statusCode = [(NSHTTPURLResponse *)response statusCode];
        BOOL success = !error && statusCode >= 200 && statusCode < 300;
        dispatch_async(dispatch_get_main_queue(), ^{
            completion(success);
        });
    }];
    [task resume];
}

@end

File: legacy_ios/LegacyApp/Sources/ConfirmFlutterViewController.swift
import Flutter
import FlutterPluginRegistrant
import UIKit

// 確認内容の送信(POST)はFlutter側(legacyapp_flutter)に寄せたため、ここで
// 実装するのはBaseViewControllerが保持する状態の受け渡しと、完了画面への
// 遷移依頼のみ。Flutter統合の新規コードはSwiftで書く方針にしたため、
// 移行前の他のViewController(Objective-C)とは異なりこのクラスだけSwift。
//
// エンジンは訪問のたびに新規生成する(アプリ全体で使い回さない)。Dartの
// main()/initState()はエンジンの生存期間中に一度しか実行されないため、
// 使い回すと2回目以降の訪問で最新の入力内容が反映されなくなる
// (詳細はdocs/MIGRATION_GUIDE.md 5節を参照)。
class ConfirmFlutterViewController: FlutterViewController {

    private static let channelName = "com.example.legacyapp/confirm"

    private let ownedEngine: FlutterEngine
    private var channel: FlutterMethodChannel!

    init() {
        let engine = FlutterEngine(name: "confirm_engine")
        // engine.binaryMessenger はエンジンがrunした後でないと使えないため、
        // run()を先に済ませてからsuper.init(engine:)でアタッチする。
        // sqfliteのようなネイティブプラグインはregister(with:)の中でも
        // engine.binaryMessenger経由のチャンネル登録を行うため、
        // GeneratedPluginRegistrant.register(with:)もrun()より後に呼ぶ
        // 必要がある(先に呼ぶと「Setting a message handler before the
        // FlutterEngine has been run」でクラッシュする)。
        engine.run()
        GeneratedPluginRegistrant.register(with: engine)

        ownedEngine = engine
        super.init(engine: engine, nibName: nil, bundle: nil)

        channel = FlutterMethodChannel(
            name: Self.channelName,
            binaryMessenger: engine.binaryMessenger
        )
        channel.setMethodCallHandler { [weak self] call, result in
            guard let self else { return }
            switch call.method {
            case "getInitialData":
                let data = BaseViewController.sharedFormData
                result(["name": data.name, "email": data.email, "message": data.message])
            case "goToComplete":
                let completeVC = CompleteViewController()
                self.navigationController?.pushViewController(completeVC, animated: true)
                result(nil)
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }

    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}
